src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -108,26 +108,22 @@ internal override void OnBufferAllocated(byte[] buffer)
         {
             Debug.Assert(_preallocatedOverlapped == null);
 
-            _preallocatedOverlapped = new PreAllocatedOverlapped(FileStreamCompletionSource.s_ioCallback, this, buffer);
+            _preallocatedOverlapped = new PreAllocatedOverlapped(FileStreamValueTaskSource.s_ioCallback, this, buffer);
         }
 
-        SafeFileHandle IFileStreamCompletionSourceStrategy.FileHandle => _fileHandle;
-
-        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CurrentOverlappedOwner => _currentOverlappedOwner;
-
-        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CompareExchangeCurrentOverlappedOwner(FileStreamCompletionSource? newSource, FileStreamCompletionSource? existingSource)
+        private FileStreamValueTaskSource? CompareExchangeCurrentOverlappedOwner(FileStreamValueTaskSource? newSource, FileStreamValueTaskSource? existingSource)
             => Interlocked.CompareExchange(ref _currentOverlappedOwner, newSource, existingSource);
 
         public override int Read(byte[] buffer, int offset, int count)
-            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).GetAwaiter().GetResult();
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).AsTask().GetAwaiter().GetResult();
```

Comment:
This will unnecessarily allocate if the operation has already completed.
```suggestion
            {
                ValueTask<int> vt = ReadAsyncInternal(new Memory<byte>(buffer, offset, count));
                return vt.Completed ?
                    vt.Result :
                    vt.AsTask().GetAwaiter().GetResult();
            }
```

Reply:
> This will unnecessarily allocate if the operation has already completed.

@stephentoub you are right, but this will happen only for reads that failed immediately and in general we don't care for code paths that throw?

FWIW @Jozkee has performed an experiment to ensure that we can't have immediate returns for happy paths: https://github.com/dotnet/runtime/issues/50087

Reply:
It doesn't require synchronous completion.  The operation could complete asynchronously fast enough that the task has already completed by the time you get here.

Reply:
> The operation could complete asynchronously fast enough that the task has already completed by the time you get here.

You are right. I hope that it happens in real life ;)